pulsar-broker-common/src/main/java/org/apache/pulsar/common/configuration/VipStatus.java

```diff
@@ -38,25 +44,75 @@ public class VipStatus {
     public static final String ATTRIBUTE_STATUS_FILE_PATH = "statusFilePath";
     public static final String ATTRIBUTE_IS_READY_PROBE = "isReadyProbe";
 
+    // log a full thread dump when a deadlock is detected in status check once every 10 minutes
+    // to prevent excessive logging
+    private static final long LOG_THREADDUMP_INTERVAL_WHEN_DEADLOCK_DETECTED = 600000L;
+    private static volatile long lastCheckStatusTimestamp;
+
+    // Since the status endpoint doesn't have authentication, it will be necessary to have a solution to prevent
+    // introducing a new DoS vulnerability where calling the status endpoint in a tight loop could introduce
+    // significant load to the system. One way would be to check that the deadlock check is executed only
+    // when there's more than 1 seconds from the previous check.
+    // If it's less than that, the previous result of the deadlock check would be reused.
+    private static final long DEADLOCK_DETECTED_INTERVAL = 1000L;
+    private static volatile boolean lastCheckStatusResult = true;
+
     @Context
     protected ServletContext servletContext;
 
     @GET
     public String checkStatus() {
-        String statusFilePath = (String) servletContext.getAttribute(ATTRIBUTE_STATUS_FILE_PATH);
-        @SuppressWarnings("unchecked")
-        Supplier<Boolean> isReadyProbe = (Supplier<Boolean>) servletContext.getAttribute(ATTRIBUTE_IS_READY_PROBE);
+        // Locking classes to avoid deadlock detection in multi-thread concurrent requests.
+        synchronized (VipStatus.class) {
```

Comment:
since the fields should be instance fields, simply make the method `synchronized` or use a limited `synchronized(this)` block as there was before.

Reply:
I tested locally and printed the current object reference every time the `checkStatus () `method was called, and found that a new object was created each time. And it was found that if the `lastCheckStatusTimestamp ` field is not decorated with `static`, it is initialized to 0 every time the` checkStatus ()` method is called, which does not match our expected target of executing the detection at a specified interval.

The printed log is shown below:
```
org.apache.pulsar.common.configuration.VipStatus@65403bed
2024-11-27T11:17:15,947 - INFO  - [pulsar-web-41-3:Slf4jRequestLogWriter@62] - [0:0:0:0:0:0:0:1] - - [27/11月/2024:11:17:15 +0800] "GET /status.html HTTP/1.1" 200 2 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36" 267
org.apache.pulsar.common.configuration.VipStatus@d560619
2024-11-27T11:17:24,500 - INFO  - [pulsar-web-41-9:Slf4jRequestLogWriter@62] - [0:0:0:0:0:0:0:1] - - [27/11月/2024:11:17:24 +0800] "GET /status.html HTTP/1.1" 200 2 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36" 6
org.apache.pulsar.common.configuration.VipStatus@646ada8f
2024-11-27T11:17:57,562 - INFO  - [pulsar-web-41-8:Slf4jRequestLogWriter@62] - [0:0:0:0:0:0:0:1] - - [27/11月/2024:11:17:57 +0800] "GET /status.html HTTP/1.1" 200 2 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36" 3

```

So, I suggest using the `static ` keyword to decorated the `lastCheckStatusTimestamp` field. Use `synchronized (VipStatus.class)` for global synchronization to avoid DoS attacks caused by multi-threaded concurrent access.